### ISSUE TITLE
Macro - Augment class with generic

### ISSUE THREAD

**Issue by LeadcodeDev** [dart-lang/language#3922]
Hi, I'm trying to apply a macro to a class that has a generic.

To experiment a little with the meta-programming approach, I set myself the goal of creating a mini ORM based on models such as Symfony or Adonis.

I have one model, that represent database table in Dart context.
 
```dart
@Model()
final class MyModel implements BaseModel {
  @Column()
  external String _firstname;

  @Column()
  external String _lastname;
}
```
The macro `@Model()` provides a static function for obtaining an instance of my query builder.

```dart
macro class QueryBuilderMacro implements ClassDeclarationsMacro {
  const QueryBuilderMacro();

  @override
  FutureOr<void> buildDeclarationsForClass(ClassDeclaration clazz, MemberDeclarationBuilder builder) async {
    final fields = await builder.fieldsOf(clazz);
    print(clazz.typeParameters.map((e) => e.name));

    final availableFields = fields
        .where((element) => !element.identifier.name.startsWith("_"))
        .toList();

    final List<Object> parts = [];
    for (final field in availableFields) {
      final paramName = field.identifier.name.replaceAll('_', '')[0].toUpperCase() + field.identifier.name.substring(1);
      final type = field.type.code.parts.first as Identifier;
      parts.addAll([
        'QueryBuilder<${clazz.identifier.name}> where$paramName($type value) {',
          'return this.where("$field", value);',
        '}',
      ]);
    }

    builder.declareInType(DeclarationCode.fromParts(parts));
  }
}
```

```dart
@QueryBuilderMacro()
final class QueryBuilder<T extends BaseModel> {
}
```

Then my main
```dart
Future<void> main(List<String> arguments) async {
  final myModel = await MyModel
    .query()
    .firstOrFail();

  print([myModel.firstname, myModel.lastname]); // prints [John, Doe]
}
```

I have this error in my console when I run my application :eyes: 
```
macros ../sdk/dart-sdk/bin/dart run --enable-experiment=macros bin/main.dart
org-dartlang-augmentation:/Development/pocs/macros/lib/query_builder.dart-0:3:21: Error: A patch class must have the same number of type variables as its origin class.
augment final class QueryBuilder {
                    ^
lib/query_builder.dart:6:13: Context: This is the origin class.
final class QueryBuilder<T extends BaseModel> {
            ^
```

I'd like to know whether this problem is actually a limitation to the use of macros or whether I'm doing it wrong. If so, could you give me some alternatives?

**Comment by davidmorgan** [comment]
I think getting the generics right here is simply not implemented yet.

See also #3879 for discussion as to whether it might be useful augmentation libraries were less strict here.

**Comment by LeadcodeDev** [comment]
Hello @davidmorgan, thank you for your reply!
I'll be following this thread closely 🙂

**Comment by jakemac53** [comment]
Yeah this just seems like a bug in the code producing the augmentation for types with generics.

**Comment by jakemac53** [comment]
Fix out here https://dart-review.googlesource.com/c/sdk/+/372580

**Comment by LeadcodeDev** [comment]
Thanks you so much 🙏

**Comment by jakemac53** [comment]
This should be resolved, but you may need to wait a bit for a dev SDK to be released that includes the fix (needs to be >= 3.5.0-288.0.dev)

**Comment by LeadcodeDev** [comment]
No problem, thank you.
Thank you for your efforts, keep up the good work !
